### PR TITLE
fix(gitlab): verify if MR is mergeable before setting automerge

### LINKED ISSUE #6248: Gitlab: "405 method not allowed" when enabling gitlabAutomerge
**What Renovate type are you using?**

Renovate OpenSource CLI on Gitlab.com pipeline (can reproduce locally though).

**Describe the bug**

Gitlab automerge does not work consistently when `gitlabAutomerge` is set. Most of the time I get a `405 method not allowed`.

**Did you see anything helpful in debug logs?**

<details>
<summary>Expand for log</summary>

```
DEBUG: Creating PR (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "prTitle": "build(deps): update dependency apollo-client to v2.6.10"
DEBUG: Creating Merge Request: build(deps): update dependency apollo-client to v2.6.10 (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
DEBUG: Gitlab API error (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "err": {
         "name": "HTTPError",
         "hostname": "gitlab.com",
         "method": "PUT",
         "path": "/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
         "protocol": "https:",
         "url": "https://gitlab.com/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
         "gotOptions": {
           "path": "/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
           "protocol": "https:",
           "hostname": "gitlab.com",
           "hash": "",
           "search": "",
           "pathname": "/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
           "href": "https://gitlab.com/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
           "headers": {
             "user-agent": "https://github.com/renovatebot/renovate",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate",
             "content-type": "application/json",
             "Private-token": "**redacted**",
             "content-length": 72
           },
           "hooks": {
             "beforeError": [],
             "init": [],
             "beforeRequest": [],
             "beforeRedirect": [],
             "beforeRetry": [],
             "afterResponse": []
           },
           "retry": {"methods": {}, "statusCodes": {}, "errorCodes": {}},
           "decompress": true,
           "throwHttpErrors": true,
           "followRedirect": true,
           "stream": false,
           "form": false,
           "json": true,
           "cache": false,
           "useElectronNet": false,
           "hostType": "gitlab",
           "baseUrl": "https://gitlab.com/api/v4/",
           "body": "{\"should_remove_source_branch\":true,\"merge_when_pipeline_succeeds\":true}",
           "method": "PUT",
           "gotTimeout": {"request": 60000}
         },
         "statusCode": 405,
         "statusMessage": "Method Not Allowed",
         "headers": {
           "date": "Fri, 15 May 2020 09:16:17 GMT",
           "content-type": "application/json",
           "content-length": "36",
           "connection": "close",
           "set-cookie": [
             "__cfduid=daae1f48e99eb6797bf659521a62876151589534177; expires=Sun, 14-Jun-20 09:16:17 GMT; path=/; domain=.gitlab.com; HttpOnly; SameSite=Lax; Secure"
           ],
           "cache-control": "no-cache",
           "vary": "Origin",
           "x-content-type-options": "nosniff",
           "x-frame-options": "SAMEORIGIN",
           "x-request-id": "SxkwjNLnwV2",
           "x-runtime": "0.124931",
           "ratelimit-limit": "600",
           "ratelimit-observed": "29",
           "ratelimit-remaining": "571",
           "ratelimit-reset": "1589534237",
           "ratelimit-resettime": "Fri, 15 May 2020 09:17:17 GMT",
           "gitlab-lb": "fe-01-lb-gprd",
           "gitlab-sv": "localhost",
           "cf-cache-status": "DYNAMIC",
           "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
           "server": "cloudflare",
           "cf-ray": "593bc25e5aac0bed-AMS",
           "cf-request-id": "02b937cef500000bedd534f200000001"
         },
         "body": {"message": "405 Method Not Allowed"},
         "message": "Response code 405 (Method Not Allowed)",
         "stack": "HTTPError: Response code 405 (Method Not Allowed)\n    at EventEmitter.<anonymous> (/home/tmeijn/_COMPANY_/tools/renovate-bot/node_modules/got/source/as-promise.js:74:19)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
       }
DEBUG: Automerge on PR creation failed (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "err": {
         "name": "HTTPError",
         "hostname": "gitlab.com",
         "method": "PUT",
         "path": "/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
         "protocol": "https:",
         "url": "https://gitlab.com/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
         "gotOptions": {
           "path": "/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
           "protocol": "https:",
           "hostname": "gitlab.com",
           "hash": "",
           "search": "",
           "pathname": "/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
           "href": "https://gitlab.com/api/v4/projects/_COMPANY_%2F_PROJECT_%2Fservices%2Fclient/merge_requests/815/merge",
           "headers": {
             "user-agent": "https://github.com/renovatebot/renovate",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate",
             "content-type": "application/json",
             "Private-token": "**redacted**",
             "content-length": 72
           },
           "hooks": {
             "beforeError": [],
             "init": [],
             "beforeRequest": [],
             "beforeRedirect": [],
             "beforeRetry": [],
             "afterResponse": []
           },
           "retry": {"methods": {}, "statusCodes": {}, "errorCodes": {}},
           "decompress": true,
           "throwHttpErrors": true,
           "followRedirect": true,
           "stream": false,
           "form": false,
           "json": true,
           "cache": false,
           "useElectronNet": false,
           "hostType": "gitlab",
           "baseUrl": "https://gitlab.com/api/v4/",
           "body": "{\"should_remove_source_branch\":true,\"merge_when_pipeline_succeeds\":true}",
           "method": "PUT",
           "gotTimeout": {"request": 60000}
         },
         "statusCode": 405,
         "statusMessage": "Method Not Allowed",
         "headers": {
           "date": "Fri, 15 May 2020 09:16:17 GMT",
           "content-type": "application/json",
           "content-length": "36",
           "connection": "close",
           "set-cookie": [
             "__cfduid=daae1f48e99eb6797bf659521a62876151589534177; expires=Sun, 14-Jun-20 09:16:17 GMT; path=/; domain=.gitlab.com; HttpOnly; SameSite=Lax; Secure"
           ],
           "cache-control": "no-cache",
           "vary": "Origin",
           "x-content-type-options": "nosniff",
           "x-frame-options": "SAMEORIGIN",
           "x-request-id": "SxkwjNLnwV2",
           "x-runtime": "0.124931",
           "ratelimit-limit": "600",
           "ratelimit-observed": "29",
           "ratelimit-remaining": "571",
           "ratelimit-reset": "1589534237",
           "ratelimit-resettime": "Fri, 15 May 2020 09:17:17 GMT",
           "gitlab-lb": "fe-01-lb-gprd",
           "gitlab-sv": "localhost",
           "cf-cache-status": "DYNAMIC",
           "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
           "server": "cloudflare",
           "cf-ray": "593bc25e5aac0bed-AMS",
           "cf-request-id": "02b937cef500000bedd534f200000001"
         },
         "body": {"message": "405 Method Not Allowed"},
         "message": "Response code 405 (Method Not Allowed)",
         "stack": "HTTPError: Response code 405 (Method Not Allowed)\n    at EventEmitter.<anonymous> (/home/tmeijn/_COMPANY_/tools/renovate-bot/node_modules/got/source/as-promise.js:74:19)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
       }
 INFO: PR created (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "pr": 815,
       "prTitle": "build(deps): update dependency apollo-client to v2.6.10"
DEBUG: getBranchStatus(renovate/patch-apollo-graphql-packages) (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
DEBUG: Got res with 45 results (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
DEBUG: getBranchStatus() result (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "branchStatus": "red",
       "branchName": "renovate/patch-apollo-graphql-packages"
DEBUG: Adding assignees tmeijn to #815 (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
DEBUG: Added assignees (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "assignees": ["tmeijn"]
DEBUG: Created Merge Request #815 (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
DEBUG: Checking #815 for automerge (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "automerge": true,
       "automergeType": "pr",
       "automergeComment": "automergeComment"
DEBUG: PR is configured for automerge (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
DEBUG: PR is not ready for merge (repository=_COMPANY_/_PROJECT_/services/client, branch=renovate/patch-apollo-graphql-packages)
       "canMergeReason": undefined
```

</details>

**To Reproduce**

This one is difficult, because the behavior is not consistent, but basically:

1. Have a dependency that fulfills your automerge conditions fullfilled.
2. Enable gitlabAutomerge in the config
3. For me, Renovate will return most of the time `405: Method not allowed`

**Additional context**

The gitlabAutomerge feature calls the Gitlab API `:merge_request/merge` endpoint with `merge_when_pipeline_succeeds: true`. According to the [API doc](https://docs.gitlab.com/ee/api/merge_requests.html#accept-mr) it will return a `405` if "merge request is unable to be accepted (ie: Work in Progress, Closed, Pipeline Pending Completion, or Failed while requiring Success)".

According to a note in the [API Doc for getting a single merge request](https://docs.gitlab.com/ee/api/merge_requests.html#get-single-mr) the mergability of a merge request is now checked asynchronously. This feature flag is now enabled by default on Gitlab.com ([source](https://gitlab.com/gitlab-org/gitlab/-/issues/199912#note_337958125)). 

my reasoning therefore is that trying to set `merge_when_pipeline_succeeds` immediately results in an error because the merge request status is not `can_be_merged` when trying to do so.

_NOTE: after trying to patch the code to fix this behavior I also found an issue when there is no pipeline created for the merge request yet so I also added the predicate that a pipeline must exist for the merge request._

This suspicion is further corroborated by patching the relevant code like this (I did it directly in the node_modules package):

```javascript
if (platformOptions && platformOptions.gitLabAutomerge) {
        function sleep(ms) {
            return new Promise(resolve => setTimeout(resolve, ms));
        }
        try {
            const desiredStatus = 'can_be_merged';
            const retryTimes = 5;

            for (let attempt = 1; attempt <= retryTimes; attempt++) {
                console.log('DEBUGGING HERE', attempt);
                const { body } = await gl_got_wrapper_1.api.get(`projects/${config.repository}/merge_requests/${pr.iid}`);
                console.log('DEBUGGING HERE', body.merge_status, body.pipeline);
                // Only continue if the merge request and the merge request has a pipeline.
                if (body.merge_status === desiredStatus && body.pipeline !== null) {
                    break;
                }
                await sleep(500 * attempt);
            }
            await gl_got_wrapper_1.api.put(`projects/${config.repository}/merge_requests/${pr.iid}/merge`, {
                body: {
                    should_remove_source_branch: true,
                    merge_when_pipeline_succeeds: true,
                },
            });
        }
        catch (err) /* istanbul ignore next */ {
            logger_1.logger.debug({ err }, 'Automerge on PR creation failed');
        }
    }
```

Solves my problem. I let Renovate create the same MR a couple of times and they all had Merge when pipeline succeeds enabled.

### PR DESCRIPTION
**Pull request by tmeijn**
This polls the Gitlab Merge requests API for the required status that's needed before setting `merge_when_pipeline_succeeds` to `true`.

Closes #6248

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
You need to mock `delay` in tests: `jest.mock('delay', () => Promise.resolve());`

**Comment by tyronedd** [comment]
@viceice Sorry I have little to no experience with testing. I tried to do what you suggested, but it still seems to fail. Can you help me out?

**Comment by viceice** [comment]
ok, can you first fix your commits to use the right author

**Comment by viceice** [comment]
You need to add the following after `describe('createPr(branchName, title, body)', () => {`.

line ~956

```js
    beforeEach(() => {
      api.get.mockResolvedValueOnce(
        partial<GotResponse>({
          body: {
            merge_status: 'can_be_merged',
            pipeline: {
              id: 29626725,
              sha: '2be7ddb704c7b6b83732fdd5b9f09d5a397b5f8f',
              ref: 'patch-28',
              status: 'success',
            },
          },
        })
      );
    });
```

**Comment by tyronedd** [comment]
@viceice @rarkins Any pointers on how to proceed? It fails on code coverage, saying the `await delay` statement is not covered. Again, no experience with testing.

File: lib/platform/gitlab/index.spec.ts [hunk 1]
```diff
@@ -952,6 +953,22 @@ describe('platform/gitlab', () => {
     });
   });
   describe('createPr(branchName, title, body)', () => {
+    beforeEach(() => {
+      api.get.mockResolvedValueOnce(
```
Comment (viceice) [inline lib/platform/gitlab/index.spec.ts:957]:
```suggestion
      api.get.mockResolvedValueOnce({ body: {} });
      api.get.mockResolvedValueOnce(
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/platform/gitlab/index.spec.ts [hunk 2]
```diff
@@ -1058,6 +1059,23 @@ describe('platform/gitlab', () => {
...
+              id: 29626725,
+              sha: '2be7ddb704c7b6b83732fdd5b9f09d5a397b5f8f',
+              ref: 'patch-28',
+              status: 'success',
+            },
+          },
+        })
+      );
+    });
+
```
Comment (viceice) [inline lib/platform/gitlab/index.spec.ts]:
```suggestion
```

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
Why is this change necessary? It will slow down automerging for GitLab users.

**Comment by tyronedd** [comment]
> Do you expect that in most cases the flag will still be set in the first run, or in a later run? Have you verified that the logic will still work on a later run?

So I'm on Renovate 19.76.3 and of these MRs:

image

Log of the relevant job can by found here: https://pastebin.com/YHaZXhW7

all of them except protractor should've been Gitlab automerged by setting merge when pipeline succeeds. They were not however, they were merged by a second run of Renovate bot. You can see in the log they all returned 405 when trying to set the flag.

If you mean a later run of Renovate bot, I do not think Renovate as of 19.76.3 tries to set the merge when pipeline succeeds flag when a merge request is already created in a previous run.

Difficult thing is, I tried to reason about this as thorough as I could by going through the docs and issues in Gitlab, but I didn't hear anybody else here having this issue.

**Pull request merged**